beluga/include/beluga/circular_array.hpp

```diff
@@ -0,0 +1,539 @@
+#include <stdexcept>
+#include <type_traits>
+#include <utility>
+
+#include <beluga/utility/indexing_iterator.hpp>
+
+/**
+ * \file
+ * \brief Implementation of a generic circular array container.
+ */
+
+namespace beluga {
+
+/// Feature flags for circular arrays.
+enum class CircularArrayFeatureFlags : int {
+  kNone = 0x00,
+  kRolloverOnWrite = 0x01,    ///<! If enabled, older values in the array are overwritten by
+                              /// newer values if the array has reached its maximum size already.
+  kExtrapolateOnRead = 0x02,  ///<! If enabled, the back value is extrapolated for constant
+                              /// accesses up to the array maximum size.
+  kLayoutReversal = 0x04      ///<! If enabled, the circular array memory layout is reversed so that
+                              /// values can be pushed to the front rather than the back of the array.
+};
+
+/// Bitwise OR operator overload to combine two feature flags in a single mask-like flag.
+inline constexpr CircularArrayFeatureFlags operator|(CircularArrayFeatureFlags lflag, CircularArrayFeatureFlags rflag) {
+  return static_cast<CircularArrayFeatureFlags>(
+      static_cast<typename std::underlying_type<CircularArrayFeatureFlags>::type>(lflag) |
+      static_cast<typename std::underlying_type<CircularArrayFeatureFlags>::type>(rflag));
+}
+
+/// Bitwise AND operator overload to check of the presence of a feature `flag` in a feature `mask`.
+inline constexpr bool operator&(CircularArrayFeatureFlags mask, CircularArrayFeatureFlags flag) {
+  return static_cast<bool>(
+      static_cast<typename std::underlying_type<CircularArrayFeatureFlags>::type>(mask) &
+      static_cast<typename std::underlying_type<CircularArrayFeatureFlags>::type>(flag));
+}
+
+/// An implementation of generic, non-threadsafe circular array.
+/**
+ * Modelled after `std::array`. Its maximum size is fixed at compile-time. Most operations are O(1)
+ * but for a few that exhibit worse case O(N) (not considering the complexity of move-construction
+ * and move-assignment of array elements of type T). Additionally, it features optional indexing
+ * reversal (making it a LIFO rather than a FIFO data structure), optional back value extrapolation
+ * for constant read accesses, and optional rollover on write accesses (i.e. overwriting the oldest
+ * value when pushing to an array instance that is full).
+ *
+ * \tparam T Element type.
+ *   It must be [default constructible](https://en.cppreference.com/w/cpp/named_req/DefaultConstructible),
+ *   [move constructible](https://en.cppreference.com/w/cpp/named_req/MoveConstructible), and
+ *   [move assignable](https://en.cppreference.com/w/cpp/named_req/MoveAssignable).
+ * \tparam N Maximum size. It must be a positive integer.
+ * \tparam F Array feature flags, to enable optional functionality. It defaults to none.
+ */
+template <typename T, std::size_t N, CircularArrayFeatureFlags F = CircularArrayFeatureFlags::kNone>
+class CircularArray {
```

Comment:
Funnily enough, this needs to define an `allocator_type` for `ranges::to` to work in range-v3 10.0:

https://github.com/ericniebler/range-v3/blob/0.10.0/include/range/v3/range/conversion.hpp#L262

Maybe add `using allocator_type = void;` somewhere down below.

Reply:
Haha, I love the `HACKHACK` comment :joy:

Reply:
Done in https://github.com/Ekumen-OS/beluga/commit/d451b5add74832b2397a3625c1a58d4565b409eb.